packages/library/src/commonMain/kotlin/io/realm/internal/RealmReference.kt

```diff
@@ -0,0 +1,37 @@
+package io.realm.internal
+
+import io.realm.BaseRealm
+import io.realm.interop.NativePointer
+
+/**
+ * FIXME Update
+ * A TransactionId makes it possible to track which public Realm instance an underlying C++ SharedRealm is associated
+ * with (represented by the `dbPointer`). This is needed as each Results, List or Object need to know, both which
+ * public Realm they belong to, but also what underlying SharedRealm they are part of. But since the public Realm
+ * instance might update its own `dbPointer`, derived objects cannot just keep a pointer to the public Realm instance.
+ *
+ * For frozen Realms, the `dbPointer` will thus point to a specific version of a read transaction that is guaranteed
+ * to not change.
+ *
+ * For live Realms, the `dbPointer` will point to a live SharedRealm that can advance its internal version.
+ *
+ * For both versions, the public Realm reference is allowed to change independently from the underlying
+ * `dbPointer`, so care should be taken when accessing any methods on the public Realm.
+ */
+data class RealmReference(
+    val owner: BaseRealm,
+    val dbPointer: NativePointer
+    // FIXME Should we keep a debug flag to assert that we have the right liveness state
+)
+
+internal interface RealmReferrer {
```

Comment:
It looks problematic to mix internal and external methods/properties in this interface. I.e. `realm` is internal, but the rest is not. But it is rather unfortunate that it would have made it quite easy to add default implementations to classes.

Not sure if we can have a public interface extend another internal interface, but I cannot come up with a name that would make sense ... `RealmLifeCycle`?

Also, `RealmReferrer` sounds a bit weird to me, I guess we have a reference to it, but it still feels off somehow. Not sure I have a better name right now though

Reply:
I have introduced internal `RealmLifeCycle` interface to hold the common interface of the state of any objects tied to a Realm. The interface is currently only implemented for the internal `RealmReference`, but could also be used to do a common default implementation across that could be used across all objects holding a `RealmReference`. But, if we expose this on any public stuff (which I think would indeed make sense), we would need to break up `Realm`, `BaseRealm`, etc. up into public interfaces and internal implementations (which I think would indeed make sense).  But left for another PR.